File: python/tests/test_encrypted_transport.py
"""v1.0 encrypted transport round-trip tests.

Covers the full loop: plaintext → encrypted .mpgo → transport
stream → decrypted .mpgo → decrypt plaintext, and variants with
encrypted AU headers.
"""
from __future__ import annotations

import io
from pathlib import Path

import numpy as np
import pytest

pytest.importorskip("cryptography")
pytest.importorskip("h5py")

from mpeg_o.enums import AcquisitionMode, Polarity
from mpeg_o.encryption_per_au import decrypt_per_au_file, encrypt_per_au_file
from mpeg_o.feature_flags import (
    OPT_ENCRYPTED_AU_HEADERS,
    OPT_PER_AU_ENCRYPTION,
)
from mpeg_o.spectral_dataset import SpectralDataset, WrittenRun
from mpeg_o.transport.codec import TransportReader, TransportWriter
from mpeg_o.transport.encrypted import (
    is_per_au_encrypted,
    read_encrypted_to_file,
    write_encrypted_dataset,
)
from mpeg_o.transport.packets import PacketFlag, PacketType


KEY = b"\xCD" * 32


def _make_plaintext(path: Path, n_spectra: int = 5,
                     points_per_spectrum: int = 4) -> Path:
    total = n_spectra * points_per_spectrum
    mz = np.arange(total, dtype="<f8") + 100.0
    intensity = (np.arange(total, dtype="<f8") + 1.0) * 10.0
    offsets = np.arange(0, total, points_per_spectrum, dtype="<u8")
    lengths = np.full(n_spectra, points_per_spectrum, dtype="<u4")
    rts = np.arange(n_spectra, dtype="<f8") * 1.5
    ms_levels = np.array(
        [1 if i % 2 == 0 else 2 for i in range(n_spectra)], dtype="<i4"
    )
    polarities = np.full(n_spectra, int(Polarity.POSITIVE), dtype="<i4")
    precursor_mzs = np.array(
        [0.0 if ms_levels[i] == 1 else 500.0 + i for i in range(n_spectra)],
        dtype="<f8",
    )
    precursor_charges = np.array(
        [0 if ms_levels[i] == 1 else 2 for i in range(n_spectra)], dtype="<i4"
    )
    base_peak = np.array(
        [float(intensity[i * points_per_spectrum:(i + 1) * points_per_spectrum].max())
         for i in range(n_spectra)],
        dtype="<f8",
    )
    run = WrittenRun(
        spectrum_class="MPGOMassSpectrum",
        acquisition_mode=int(AcquisitionMode.MS1_DDA),
        channel_data={"mz": mz, "intensity": intensity},
        offsets=offsets,
        lengths=lengths,
        retention_times=rts,
        ms_levels=ms_levels,
        polarities=polarities,
        precursor_mzs=precursor_mzs,
        precursor_charges=precursor_charges,
        base_peak_intensities=base_peak,
    )
    SpectralDataset.write_minimal(
        path,
        title="encrypted transport fixture",
        isa_investigation_id="ISA-ENCTRANS",
        runs={"run_0001": run},
    )
    return path


class TestDetection:

    def test_is_per_au_encrypted_before_encrypt(self, tmp_path):
        p = _make_plaintext(tmp_path / "src.mpgo")
        assert not is_per_au_encrypted(str(p))

    def test_is_per_au_encrypted_after_encrypt(self, tmp_path):
        p = _make_plaintext(tmp_path / "src.mpgo")
        encrypt_per_au_file(str(p), KEY)
        assert is_per_au_encrypted(str(p))


class TestEncryptedChannelRoundTrip:
    """ENCRYPTED-only mode (plaintext filter header)."""

    def _encrypt_then_stream(self, tmp_path: Path):
        src = _make_plaintext(tmp_path / "src.mpgo")
        encrypt_per_au_file(str(src), KEY)
        stream = io.BytesIO()
        with TransportWriter(stream) as tw:
            write_encrypted_dataset(tw, str(src))
        stream.seek(0)
        return src, stream

    def test_stream_contains_protection_metadata(self, tmp_path):
        _, stream = self._encrypt_then_stream(tmp_path)
        packet_types = []
        with TransportReader(stream) as tr:
            for header, _ in tr.iter_packets():
                packet_types.append(header.packet_type)
        assert int(PacketType.PROTECTION_METADATA) in packet_types

    def test_encrypted_flag_on_aus(self, tmp_path):
        _, stream = self._encrypt_then_stream(tmp_path)
        with TransportReader(stream) as tr:
            for header, _ in tr.iter_packets():
                if header.packet_type == int(PacketType.ACCESS_UNIT):
                    assert header.flags & int(PacketFlag.ENCRYPTED)
                    assert not (header.flags & int(PacketFlag.ENCRYPTED_HEADER))

    def test_full_roundtrip(self, tmp_path):
        src, stream = self._encrypt_then_stream(tmp_path)
        rt_path = tmp_path / "rt.mpgo"
        meta = read_encrypted_to_file(stream, rt_path)
        assert meta["title"] == "encrypted transport fixture"
        assert meta["runs"]["run_0001"]["n_spectra"] == 5
        # Receiver-side file is encrypted too.
        assert is_per_au_encrypted(str(rt_path))

        # Decrypt both ends; signal values must match.
        originals = decrypt_per_au_file(str(src), KEY)["run_0001"]
        rt_values = decrypt_per_au_file(str(rt_path), KEY)["run_0001"]
        for cname in ("mz", "intensity"):
            np.testing.assert_array_equal(originals[cname], rt_values[cname])


class TestEncryptedHeaderRoundTrip:
    """ENCRYPTED | ENCRYPTED_HEADER mode."""

    def _encrypt_headers_then_stream(self, tmp_path: Path):
        src = _make_plaintext(tmp_path / "src.mpgo")
        encrypt_per_au_file(str(src), KEY, encrypt_headers=True)
        stream = io.BytesIO()
        with TransportWriter(stream) as tw:
            write_encrypted_dataset(tw, str(src))
        stream.seek(0)
        return src, stream

    def test_both_flags_set_on_aus(self, tmp_path):
        _, stream = self._encrypt_headers_then_stream(tmp_path)
        with TransportReader(stream) as tr:
            for header, _ in tr.iter_packets():
                if header.packet_type == int(PacketType.ACCESS_UNIT):
                    assert header.flags & int(PacketFlag.ENCRYPTED)
                    assert header.flags & int(PacketFlag.ENCRYPTED_HEADER)

    def test_full_roundtrip_with_headers(self, tmp_path):
        src, stream = self._encrypt_headers_then_stream(tmp_path)
        rt_path = tmp_path / "rt.mpgo"
        meta = read_encrypted_to_file(stream, rt_path)
        assert meta["runs"]["run_0001"]["encrypted_headers"] is True
        assert is_per_au_encrypted(str(rt_path))

        # Both files should carry opt_encrypted_au_headers.
        import h5py
        with h5py.File(str(rt_path), "r") as f:
            raw = f.attrs.get("mpeg_o_features", b"[]")
            if isinstance(raw, bytes):
                raw = raw.decode()
            assert OPT_ENCRYPTED_AU_HEADERS in raw

        originals = decrypt_per_au_file(str(src), KEY)["run_0001"]
        rt_values = decrypt_per_au_file(str(rt_path), KEY)["run_0001"]
        for cname in ("mz", "intensity"):
            np.testing.assert_array_equal(originals[cname], rt_values[cname])
        # Header decrypt recovers fields on the receiver side.
        assert "__au_headers__" in rt_values
        orig_headers = originals["__au_headers__"]
        rt_headers = rt_values["__au_headers__"]
        assert orig_headers == rt_headers

File: objc/Tests/TestEncryptedTransport.m
/*
 * TestEncryptedTransport — v1.0 ObjC.
 *
 * Verifies TTIOEncryptedTransport emits a well-formed stream from a
 * per-AU-encrypted file:
 *   - StreamHeader + ProtectionMetadata + DatasetHeader + N AUs +
 *     EndOfDataset + EndOfStream appear in the expected order.
 *   - Every AU packet header carries ENCRYPTED (and ENCRYPTED_HEADER
 *     when opt_encrypted_au_headers is set).
 *   - ProtectionMetadata payload is parseable (cipher_suite,
 *     kek_algorithm, wrapped_dek, etc.).
 *
 * Also verifies the reader-side path: decode a stream back into a
 * per-AU-encrypted .tio, decrypt with the same DEK, and confirm
 * plaintext channels survive byte-for-byte.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#import <Foundation/Foundation.h>
#import "Testing.h"
#import <unistd.h>
#import <string.h>

#import "Transport/TTIOEncryptedTransport.h"
#import "Transport/TTIOTransportWriter.h"
#import "Transport/TTIOTransportReader.h"
#import "Transport/TTIOTransportPacket.h"
#import "Protection/TTIOPerAUFile.h"
#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOWrittenRun.h"
#import "ValueClasses/TTIOEnums.h"


static NSString *tmpPath(NSString *n) {
    return [NSString stringWithFormat:@"/tmp/ttio_enctrans_%d_%@",
            (int)getpid(), n];
}
static void rmFile(NSString *p) { [[NSFileManager defaultManager] removeItemAtPath:p error:NULL]; }
static NSData *testKey(void) {
    uint8_t b[32]; memset(b, 0x77, 32);
    return [NSData dataWithBytes:b length:32];
}

static NSData *f64arr(const double *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}
static NSData *u64arr(const uint64_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 8];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:8];
    return d;
}
static NSData *u32arr(const uint32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}
static NSData *i32arr(const int32_t *v, NSUInteger n) {
    NSMutableData *d = [NSMutableData dataWithCapacity:n * 4];
    for (NSUInteger i = 0; i < n; i++) [d appendBytes:&v[i] length:4];
    return d;
}


static BOOL buildAndEncryptFixture(NSString *path,
                                     BOOL encryptHeaders,
                                     NSError **error)
{
    NSUInteger n = 3, p = 4, total = n * p;
    double mz[12], intensity[12];
    for (NSUInteger i = 0; i < total; i++) {
        mz[i] = 100.0 + (double)i;
        intensity[i] = (double)(i + 1) * 10.0;
    }
    uint64_t offsets[3] = {0, 4, 8};
    uint32_t lengths[3] = {4, 4, 4};
    double rts[3] = {1.0, 2.0, 3.0};
    int32_t msLevels[3] = {1, 2, 1};
    int32_t pols[3] = {1, 1, 1};
    double pmzs[3] = {0.0, 500.0, 0.0};
    int32_t pcs[3] = {0, 2, 0};
    double bpis[3] = {40.0, 80.0, 120.0};

    TTIOWrittenRun *run =
        [[TTIOWrittenRun alloc]
            initWithSpectrumClassName:@"TTIOMassSpectrum"
                      acquisitionMode:(int64_t)TTIOAcquisitionModeMS1DDA
                          channelData:@{@"mz": f64arr(mz, total),
                                        @"intensity": f64arr(intensity, total)}
                              offsets:u64arr(offsets, n)
                              lengths:u32arr(lengths, n)
                       retentionTimes:f64arr(rts, n)
                             msLevels:i32arr(msLevels, n)
                           polarities:i32arr(pols, n)
                         precursorMzs:f64arr(pmzs, n)
                     precursorCharges:i32arr(pcs, n)
                  basePeakIntensities:f64arr(bpis, n)];
    if (![TTIOSpectralDataset writeMinimalToPath:path
                                            title:@"enc-transport fixture"
                               isaInvestigationId:@"ISA-ENC-TX"
                                           msRuns:@{@"run_0001": run}
                                  identifications:nil
                                  quantifications:nil
                                provenanceRecords:nil
                                            error:error]) return NO;
    return [TTIOPerAUFile encryptFilePath:path
                                        key:testKey()
                           encryptHeaders:encryptHeaders
                              providerName:nil
                                     error:error];
}


void testEncryptedTransport(void)
{
    // ── 1. ENCRYPTED-only path ──────────────────────────────────
    {
        NSString *src = tmpPath(@"src1.tio");
        rmFile(src);
        NSError *err = nil;
        PASS(buildAndEncryptFixture(src, NO, &err),
             "encrypted fixture built (channels only)");
        PASS([TTIOEncryptedTransport isPerAUEncryptedAtPath:src
                                                providerName:nil],
             "isPerAUEncrypted reports YES after encryption");

        NSMutableData *streamBuf = [NSMutableData data];
        TTIOTransportWriter *writer =
            [[TTIOTransportWriter alloc] initWithMutableData:streamBuf];
        BOOL ok = [TTIOEncryptedTransport writeEncryptedDataset:src
                                                           writer:writer
                                                     providerName:nil
                                                            error:&err];
        [writer close];
        PASS(ok, "writeEncryptedDataset succeeds");
        PASS(streamBuf.length > 0, "stream has content");

        // Parse packets back.
        TTIOTransportReader *reader =
            [[TTIOTransportReader alloc] initWithData:streamBuf];
        NSArray *packets = [reader readAllPacketsWithError:&err];
        PASS(packets != nil, "stream parses");

        NSUInteger nStreamHeader = 0, nDatasetHeader = 0, nProtection = 0;
        NSUInteger nAU = 0, nEOD = 0, nEOS = 0;
        BOOL allAUsEncrypted = YES;
        BOOL anyHeaderEncrypted = NO;
        for (TTIOTransportPacketRecord *r in packets) {
            switch (r.header.packetType) {
                case TTIOTransportPacketStreamHeader: nStreamHeader++; break;
                case TTIOTransportPacketDatasetHeader: nDatasetHeader++; break;
                case TTIOTransportPacketProtectionMetadata: nProtection++; break;
                case TTIOTransportPacketAccessUnit:
                    nAU++;
                    if (!(r.header.flags & TTIOTransportPacketFlagEncrypted))
                        allAUsEncrypted = NO;
                    if (r.header.flags & TTIOTransportPacketFlagEncryptedHeader)
                        anyHeaderEncrypted = YES;
                    break;
                case TTIOTransportPacketEndOfDataset: nEOD++; break;
                case TTIOTransportPacketEndOfStream: nEOS++; break;
                default: break;
            }
        }
        PASS(nStreamHeader == 1, "exactly one StreamHeader");
        PASS(nDatasetHeader == 1, "exactly one DatasetHeader");
        PASS(nProtection == 1, "exactly one ProtectionMetadata");
        PASS(nAU == 3, "three AccessUnits (3-spectrum fixture)");
        PASS(nEOD == 1 && nEOS == 1, "EndOfDataset + EndOfStream present");
        PASS(allAUsEncrypted,
             "every AU carries TTIOTransportPacketFlagEncrypted");
        PASS(!anyHeaderEncrypted,
             "no AU carries EncryptedHeader (channels-only mode)");
        rmFile(src);
    }

    // ── 2. ENCRYPTED | ENCRYPTED_HEADER path ────────────────────
    {
        NSString *src = tmpPath(@"src2.tio");
        rmFile(src);
        NSError *err = nil;
        buildAndEncryptFixture(src, YES, &err);

        NSMutableData *streamBuf = [NSMutableData data];
        TTIOTransportWriter *writer =
            [[TTIOTransportWriter alloc] initWithMutableData:streamBuf];
        [TTIOEncryptedTransport writeEncryptedDataset:src
                                                 writer:writer
                                           providerName:nil
                                                  error:&err];
        [writer close];

        TTIOTransportReader *reader =
            [[TTIOTransportReader alloc] initWithData:streamBuf];
        NSArray *packets = [reader readAllPacketsWithError:&err];
        BOOL allEncrypted = YES, allEncryptedHeader = YES;
        NSUInteger nAU = 0;
        for (TTIOTransportPacketRecord *r in packets) {
            if (r.header.packetType == TTIOTransportPacketAccessUnit) {
                nAU++;
                if (!(r.header.flags & TTIOTransportPacketFlagEncrypted))
                    allEncrypted = NO;
                if (!(r.header.flags & TTIOTransportPacketFlagEncryptedHeader))
                    allEncryptedHeader = NO;
            }
        }
        PASS(nAU == 3 && allEncrypted && allEncryptedHeader,
             "all 3 AUs carry ENCRYPTED | ENCRYPTED_HEADER");
        rmFile(src);
    }

    // ── 3. Round-trip: write stream → read stream → decrypt ─────
    for (int mode = 0; mode < 2; mode++) {
        BOOL encryptHeaders = (mode == 1);
        NSString *src = tmpPath(encryptHeaders ? @"rt_src_hdr.tio"
                                               : @"rt_src_ch.tio");
        NSString *dst = tmpPath(encryptHeaders ? @"rt_dst_hdr.tio"
                                               : @"rt_dst_ch.tio");
        rmFile(src); rmFile(dst);
        NSError *err = nil;
        PASS(buildAndEncryptFixture(src, encryptHeaders, &err),
             "round-trip: encrypted fixture built");

        NSMutableData *streamBuf = [NSMutableData data];
        TTIOTransportWriter *writer =
            [[TTIOTransportWriter alloc] initWithMutableData:streamBuf];
        BOOL wrote = [TTIOEncryptedTransport writeEncryptedDataset:src
                                                              writer:writer
                                                        providerName:nil
                                                               error:&err];
        [writer close];
        PASS(wrote, "round-trip: writeEncryptedDataset");

        BOOL read = [TTIOEncryptedTransport readEncryptedToPath:dst
                                                      fromStream:streamBuf
                                                    providerName:nil
                                                           error:&err];
        PASS(read, "round-trip: readEncryptedToPath materialises file");
        PASS([TTIOEncryptedTransport isPerAUEncryptedAtPath:dst
                                                providerName:nil],
             "round-trip: output file carries opt_per_au_encryption");

        NSDictionary *srcPlain =
            [TTIOPerAUFile decryptFilePath:src key:testKey()
                              providerName:nil error:&err];
        NSDictionary *dstPlain =
            [TTIOPerAUFile decryptFilePath:dst key:testKey()
                              providerName:nil error:&err];
        PASS(srcPlain != nil && dstPlain != nil,
             "round-trip: both files decrypt");
        PASS([srcPlain[@"run_0001"][@"mz"]
                isEqualToData:dstPlain[@"run_0001"][@"mz"]],
             "round-trip: mz bytes survive transport");
        PASS([srcPlain[@"run_0001"][@"intensity"]
                isEqualToData:dstPlain[@"run_0001"][@"intensity"]],
             "round-trip: intensity bytes survive transport");

        rmFile(src); rmFile(dst);
    }
}
